
# ezASCAT

The goal of `ezASCAT` is to make life simpler while using [ASCAT](https://github.com/VanLoo-lab/ascat) with tumor-normal pairs from WGS.
Although there exists [ascatNgs](https://github.com/cancerit/ascatNgs), it requires installation of perl and C modules. `ezASCAT` bypasses these requirements entirely within R with the C code baked in.

## Installation

``` r
remotes::install_github(repo = "CompEpigen/ezASCAT")
```

## Usage

Step-1: Get nucleotide counts at the marker loci with `get_counts`

Below command will generate two tsv files `tumor_nucleotide_counts.tsv` and `normal_nucleotide_counts.tsv` that can be used for downstream analysis. Note that the function will process ~900K SNPs from [Affymetrix Genome-Wide Human SNP 6.0 Array](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GPL6801) which will take ca.15 minutes on a decent computer. Currently `hg19` and `hg38` are supported.

```r
library("ezASCAT")
counts = ezASCAT::get_counts(t_bam = "tumor.bam", n_bam = "normal.bam", build = "hg19")
```

Step-2: Prepare input files for ASCAT with `prep_ascat()`

Below command will filter SNPs with low coverage (default <30), estimate BAF, logR, and generates the input files for ASCAT.
In addition, it will run `ASCAT::ascat.loadData()` and `ASCAT::ascat.plotRawData()` for you and returns the ASCAT object that can be further processed with ASCAT functions.

```r
asc_ip = prep_ascat(t_counts = "tumor_nucleotide_counts.tsv", n_counts = "normal_nucleotide_counts.tsv", sample_name = "tumor")

# Markers: 901235
# Removed 3072 duplicated loci
# Markers > 30: 25246
# ------
# Counts file: normal_nucleotide_counts.tsv
# Markers: 901235
# Removed 3072 duplicated loci
# Markers > 30: 31387
# ------
# Final number SNPs: 23765
# Generated following files:
# tumor.tumour.BAF.txt
# tumor.tumour.logR.txt
# tumor.normal.BAF.txt
# tumor.normal.logR.txt
# ------
# Running ASCAT::ascat.loadData:
# [1] Reading Tumor LogR data...
# [1] Reading Tumor BAF data...
# [1] Reading Germline LogR data...
# [1] Reading Germline BAF data...
# [1] Registering SNP locations...
# [1] Splitting genome in distinct chunks...
# Running ASCAT::ascat.plotRawData:
# [1] Plotting tumor data
# [1] Plotting germline data
# Returned ASCAT object
```

Alternatively, logR files generated by `prep_ascat` can be processed with `segment_logR` function which performs circular binary segmentation using [DNAcopy](https://bioconductor.org/packages/release/bioc/html/DNAcopy.html) and plots the results

```r
ezASCAT::segment_logR(tumor_logR = "tumor.tumour.logR.txt", sample_name = "tumor")
```
